compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  // Big long, silly block of manual code generation.
+  gen += this->genComment(toString(fn));
+  gen += "int64_t chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "(void)";
+  gen += scope_begin;
+
+  gen += this->genServersideRPC(fn);
+  gen += "return 0;\n";
+ 
+  gen += scope_end;
+  gen += "\n";
+
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+ 
+  return;
+}
+
+std::string MLIContext::genDebugPrintCall(FnSymbol* fn, const char* pfx) {
+  std::string msg;
+
+  msg += "Calling: ";
+  msg += toString(fn);
+
+  return this->genDebugPrintCall(msg.c_str(), pfx);
+}
+
+std::string MLIContext::genDebugPrintCall(const char* msg, const char* pfx) {
+  std::string gen;
+
+  gen += "printf(\"";
+
+  if (pfx && pfx != "") {
+    gen += pfx;
+    gen += " ";
+  }
+
+  gen += "%%s\\n\", \"";
+  gen += msg;
+  gen += "\");\n";
+
+  return gen;
+}
+
+std::string MLIContext::genFuncNumericID(FnSymbol* fn) {
+  return std::to_string((int64_t) fn->id);
+}
+
+std::string MLIContext::genServerWrapperCall(FnSymbol* fn) {
+  std::string gen;
+
+  gen += "chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "();\n";
+
+  return gen;
+}
+  
+std::string
+MLIContext::genServerDispatchSwitch(const std::vector<FnSymbol*>& fns) {
+  std::string gen;
+
+  gen += "int64_t chpl_mli_sdispatch";
+  gen += "(int64_t function)";
+  gen += scope_begin;
+  gen += this->genNewDecl("int", "err");
+  gen += "switch (function)";
+  gen += scope_begin;
+
+  for_vector(FnSymbol, fn, fns) {
+    if (not fn->hasFlag(FLAG_EXPORT) || fn->hasFlag(FLAG_GEN_MAIN_FUNC)) {
+      continue;
+    }
+
+    gen += "case ";
+    gen += this->genFuncNumericID(fn);
+    gen += ": ";
+
+    gen += scope_begin;
+    
+    if (this->injectDebugPrintlines) {
+      gen += this->genDebugPrintCall(fn, "[Server]");
+    }
+    
+    gen += "err = ";
+    gen += this->genServerWrapperCall(fn);
+    gen += scope_end;
+    gen += "break;\n";
+  }
+
+  gen += "default: return CHPL_MLI_ERROR_NOFUNC; break;\n";
+  gen += scope_end;
+  gen += "return err;\n";
+  gen += scope_end;
+
+  return gen;
+}
+
+//
+// TODO: This is unused right now (no struct type support yet).
+//
+bool MLIContext::structContainsOnlyPrimitiveScalars(Type *t) {
+  return false;
+}
+
+//
+// TODO: This filter will change as we support more and more type classes.
+//  - [ ] dtStringC
+//  - [ ] Chapel `string` type
+//  - [ ] ???
+//
+bool MLIContext::isSupportedType(Type* t) {
+  return (isPrimitiveScalar(t));
+}
+
+void MLIContext::verifyPrototype(FnSymbol* fn) {
+
+  if (fn->retType != dtVoid && not isSupportedType(fn->retType)) {
+    USR_FATAL("MLI does not support code generation for type", fn->retType);
+  }
+
+  // Loop through all formals and error gen if a type is unsupported.
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    if (isSupportedType(as->type)) { continue; }
+    USR_FATAL("MLI does not support code generation for type", as->type);
```

Comment:
Do you really need a continue?  It seems like you could just flip the condition and put the USR_FATAL in the if body instead

Reply:
Ha. I have this thing where I like to avoid nested if conditions as much as possible, so this probably stems from that. I'll flip em'.